File: ltviewsx/src/main/java/com/lt/ltviewsx/lt_recyclerview/LtRecyclerViewManager.java
package com.lt.ltviewsx.lt_recyclerview;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.lt.ltviewsx.R;


/**
 * 创    建:  lt  2018/5/22--15:23
 * 作    用:  ltRv的管理者
 * 注意事项:
 */

final public class LtRecyclerViewManager {
    private Context context;//创建刷新布局时的上下文,必须有
    private Class refreshLayoutClazz = MSwipeRefreshLayout.class;//刷新布局的class
    private int upLayoutId = R.layout.lt_up_loading;//上拉加载的布局id
    private float refreshThreshold;//设置下拉的阈值
    private boolean rvIsMove = true;//RecyclerView是否跟着下拉移动
    private boolean noDataIsLoad = false;//上拉已经没数据了,再次上拉是否加载数据
    private int noItemTextColor = 0;//没条目的tv的字体颜色

    //单例
    private LtRecyclerViewManager() {
    }

    private static class LtRecyclerViewManagerInstance {
        private static final LtRecyclerViewManager mLtRecyclerViewManager = new LtRecyclerViewManager();
    }

    /**
     * 获取单实例
     */
    public static @NonNull
    LtRecyclerViewManager getInstance() {
        return LtRecyclerViewManagerInstance.mLtRecyclerViewManager;
    }

    /**
     * 初始化,在application的onCreate中调用
     */
    public @NonNull
    LtRecyclerViewManager init(@NonNull Context context) {
        this.context = context;
        this.refreshThreshold = 80 * context.getResources().getDisplayMetrics().density;
        noItemTextColor = context.getResources().getColor(R.color.color_333);
        return this;
    }

    /**
     * 获取没条目的tv的字体颜色
     */
    public int getNoItemTextColor() {
        return noItemTextColor;
    }

    /**
     * 设置没条目的tv的字体颜色
     */
    public void setNoItemTextColor(int noItemTextColor) {
        this.noItemTextColor = noItemTextColor;
    }

    /**
     * 获取实例化时传入的上下文
     */
    public @NonNull
    Context getContext() {
        return context;
    }

    /**
     * 设置下拉刷新的View的class
     */
    public @NonNull
    LtRecyclerViewManager setRefreshLayoutClazz(@NonNull Class refreshLayoutClazz) {
        this.refreshLayoutClazz = refreshLayoutClazz;
        return this;
    }

    /**
     * 获取下拉刷新的View的class
     */
    public @Nullable
    Class getRefreshLayoutClazz() {
        return refreshLayoutClazz;
    }

    /**
     * 设置上拉加载的布局id
     */
    public @NonNull
    LtRecyclerViewManager setUpLayoutId(int upLayoutId) {
        this.upLayoutId = upLayoutId;
        return this;

    }

    /**
     * 获取上拉加载的布局id
     */
    public int getUpLayoutId() {
        return upLayoutId;
    }

    /**
     * 获取下拉阈值
     */
    public float getRefreshThreshold() {
        return refreshThreshold;
    }

    /**
     * 设置下拉阈值
     */
    public @NonNull
    LtRecyclerViewManager setRefreshThreshold(float refreshThreshold) {
        this.refreshThreshold = refreshThreshold;
        return this;
    }

    /**
     * 重新设置上下文
     */
    public @NonNull
    LtRecyclerViewManager setContext(@NonNull Context context) {
        this.context = context;
        return this;
    }

    /**
     * 获取是否下拉时RecyclerView跟着向下移动
     */
    public boolean isRvIsMove() {
        return rvIsMove;
    }

    /**
     * 设置是否下拉时RecyclerView跟着向下移动
     */
    public @NonNull
    LtRecyclerViewManager setRvIsMove(boolean rvIsMove) {
        this.rvIsMove = rvIsMove;
        return this;
    }

    /**
     * 上拉已经没数据了,再次上拉是否加载数据
     */
    public boolean isNoDataIsLoad() {
        return noDataIsLoad;
    }

    /**
     * 上拉已经没数据了,再次上拉是否加载数据,默认false不加载
     */
    public @NonNull
    LtRecyclerViewManager setNoDataIsLoad(boolean noDataIsLoad) {
        this.noDataIsLoad = noDataIsLoad;
        return this;
    }
}

File: ltviewsx/src/main/java/com/lt/ltviewsx/lt_recyclerview/LtRecyclerViewManager.kt
package com.lt.ltviewsx.lt_recyclerview

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.View
import com.lt.ltviewsx.R

/**
 * 创    建:  lt  2018/5/22--15:23
 * 作    用:  ltRv的管理者
 * 注意事项:
 */
@SuppressLint("StaticFieldLeak")
object LtRecyclerViewManager {
    /**
     * 初始化,在application的onCreate中调用
     */
    fun init(context: Context): LtRecyclerViewManager {
        this.context = context
        refreshThreshold = 80 * context.resources.displayMetrics.density
        noItemTextColor = context.resources.getColor(R.color.color_333)
        return this
    }

    /**
     * 创建刷新布局时的上下文,必须有
     * 但是因为使用了Application的上下文,所以用该上下文的view无法使用配置好的主题,或设置好的density(一般上面两种情况也没什么问题),但是可以通过在Adapter的构造中传入自己new的view来避免
     */
    var context: Context? = null

    /**
     * 下拉刷新的View的构造方法引用
     */
    var refreshLayoutConstructorFunction: (Context, AttributeSet?, Int) -> BaseRefreshLayout = ::MTextRefreshLayout

    /**
     * 上拉加载的布局id
     */
    var upLayoutId = R.layout.lt_up_loading

    /**
     * 下拉阈值
     */
    var refreshThreshold = 0F

    /**
     * 下拉时RecyclerView是否跟着向下移动(仅自定义下拉刷新布局)
     */
    var isRvIsMove = true

    /**
     * 上拉已经没数据了,再次上拉是否加载数据
     */
    var isNoDataIsLoad = false

    /**
     * noItemTextView的字体颜色
     */
    var noItemTextColor = 0

    /**
     * 表示外部引用是否是Debug状态,如果是Debug状态,则不帮外部cache某些异常
     */
    var isDebug = false

    /**
     * 获取默认(自定义设置)的适配器底部刷新view
     */
    fun getDefaultBottomRefreshView(): View = View.inflate(context, upLayoutId, null)

    /**
     * Adapter的回调方法中出现了异常的处理方式,默认抛异常,用户可以自行捕获相应异常
     */
    var onAdapterCatchHandler: (Throwable) -> Unit = { throw it }
}